system/jlib/jtrace.cpp

```diff
@@ -69,3 +100,688 @@ const char* LogTrace::queryLocalId() const
+
+            ctxProps->setProp(opentelemetry::trace::propagation::kTraceState.data(), span->GetContext().trace_state()->ToHeader().c_str());
+        }
+        else
+        {
+            if (!isEmptyString(traceID.get()))
+                ctxProps->setProp("traceID", traceID.get());
+            if (!isEmptyString(spanID.get()))
+                ctxProps->setProp("spanID", spanID.get());
+            if (!isEmptyString(traceFlags.get()))
+                ctxProps->setProp("traceFlags", traceFlags.get());
+
+            if (type == SpanType::ServerSpan)
+                ctxProps->setProp("type", "Server");
+            else if (type == SpanType::ClientSpan)
+                ctxProps->setProp("type", "Client");
+            else
+                ctxProps->setProp("type", "Internal");
+
+            if (localParentSpan != nullptr)
+            {
+                Owned<IProperties> localParentSpanCtxProps = createProperties();
+                localParentSpan->getSpanContext(localParentSpanCtxProps, false);
+                if (localParentSpanCtxProps)
+                {
+                    if (localParentSpanCtxProps->hasProp("spanID"))
+                        ctxProps->setProp("localParentSpanID", localParentSpanCtxProps->queryProp("spanID"));
+                    if (localParentSpanCtxProps->hasProp("traceID"))
+                        ctxProps->setProp("localParentTraceID", localParentSpanCtxProps->queryProp("traceID"));
+                }
+            }
+
+            if (type == SpanType::ServerSpan && remoteParentSpanCtx.IsValid())
+            {
+                StringBuffer remoteParentSpanID;
+                char remoteParentSpanId[16] = {0};
+                remoteParentSpanCtx.span_id().ToLowerBase16(remoteParentSpanId);
+                remoteParentSpanID.append(16, remoteParentSpanId);
+                ctxProps->setProp("remoteParentSpanID", remoteParentSpanID.str());
+            }
+        }
+
+        if (!isEmptyString(hpccGlobalId.get()))
+            ctxProps->setProp(HPCCSemanticConventions::kGLOBALIDHTTPHeader, hpccGlobalId.get());
+
+        if (!isEmptyString(hpccCallerId.get()))
+            ctxProps->setProp(HPCCSemanticConventions::kCallerIdHTTPHeader, hpccCallerId.get());
+
+        return true;
+    }
+
+    opentelemetry::v1::trace::SpanContext querySpanContext() const
+    {
+        if (span != nullptr)
+            return span->GetContext();
+
+        return opentelemetry::trace::SpanContext::GetInvalid();
+    }
+
+    const char * queryTraceName() const
+    {
+        return tracerName.get();
+    }
+
+protected:
+    CSpan(const char * spanName, SpanType spanType, CSpan * parent)
+    {
+        name.set(spanName);
+        localParentSpan = parent;
+        if (localParentSpan != nullptr)
+            tracerName.set(parent->queryTraceName());
+        type = spanType;
+
+        init();
+    }
+
+    CSpan(const char * spanName, SpanType spanType, const char * nameOfTracer)
```

Comment:
A little odd, Serverspans use this ctr, it does not init, giving serverspans a chance to setSpanContext(httpHeaders); before calling init.
perhaps this ctr should accept httpHeaders

Reply:
Another possible pattern is for the other CSpan class to not call init(), and for a derived class (CClientSpan/CInternalSpan) to call it from the constructor.  No strong opinions unless there is another reason to create the derived classes.